File: CommonBusiness/src/main/java/com/company/commonbusiness/http/rx/BaseObserver.java
package com.company.commonbusiness.http.rx;

import com.company.commonbusiness.http.ExceptionHandle;
import com.company.commonbusiness.log.XLog;
import com.orhanobut.logger.Logger;

import io.reactivex.Observer;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;

/**
 * @author 李昭鸿
 * @desc: 观察者基类，统一处理数据及异常
 * @date Created on 2017/7/28 11:31
 */

public abstract class BaseObserver<T> implements Observer<T> {

    private Disposable disposable;

    @Override
    public void onSubscribe(@NonNull Disposable d) {
        Logger.d("BaseObserver onSubscribe");
        disposable = d;
    }

    @Override
    public void onNext(@NonNull T t) {
        Logger.d("onNext：" + t.toString());
        success(t);
    }

    @Override
    public void onError(@NonNull Throwable e) {
        error(ExceptionHandle.handleException(e).message);
    }

    @Override
    public void onComplete() {

    }

    public void dispose() {
        disposable.dispose();
    }

    protected abstract void success(T t);

    protected abstract void error(String errorMsg);
}

File: CommonBusiness/src/main/java/com/company/commonbusiness/http/rx/BaseObserver.kt
package com.company.commonbusiness.http.rx

import com.company.commonbusiness.http.ExceptionHandle
import com.orhanobut.logger.Logger

import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableObserver

/**
 * @author 李昭鸿
 * @desc: 观察者基类，统一处理数据及异常
 * @date Created on 2017/7/28 11:31
 */

abstract class BaseObserver<T> : DisposableObserver<T>() {

    private var disposable: Disposable? = null

    override fun onNext(data: T) {
        Logger.d("BaseObserver onNext: " + data.toString())
        onSuccess(data)
    }

    override fun onComplete() {
        Logger.d("BaseObserver onComplete")
    }

    override fun onError(e: Throwable) {
        Logger.d("BaseObserver onError")
        onFailure(ExceptionHandle.handleException(e).msg ?: "未知错误")
    }

    abstract fun onSuccess(data: T)

    abstract fun onFailure(errorMsg: String)
}
